### DISCUSSION TITLE
"Tag" components

### DISCUSSION THREAD

**Discussion by khalv** [skypjack/entt#651] [General]
I browsed through the documentation and noticed a mention of empty types, stating that these are not iterable through views. Having empty components (or similar functionality) is something that would be quite useful, as sometimes you just want to tag an entity and iterate over all entities with this tag (like, say, a "Static" tag). Technically, you can add a component with a single bool, but storing a value is not interesting since you just want a view into all entities with this component. This is a concept which Unity allows in their ECS system. Is there any way of doing this in entt?

**Comment by Innokentiy-Alaytsev** [comment]
Empty components are "not iterable" as in "you can't get instances of those components". So you can create a view over an empty component and iterate all the entities having this component assigned.

**Comment by skypjack** [comment]
> noticed a mention of empty types, stating that these are not iterable through views

Can I ask you where this is mentioned? Probably I should review that part and make it clear that the view is definitely iterable.
As @Innokentiy-Alaytsev said, if `T` is a non-empty type, this is an example of how you can get references to it:
```
registry.view<T>().each([](auto entity, auto &component) { ... });
```
On the other side, if `U` is an empty type, its instances is never instantiated and therefore they aren't returned. That is:
```
registry.view<T>().each([](auto entity) { ... });
```
In a multi-component view, you've that non-empty types are returned while empty types are not:
```
registry.view<T, U>().each([](auto entity, auto &t_component) { ... });
```
So, you can definitely use empty types in your views and they are treated as all other types, but for the fact that they aren't returned because they don't exist (literally!). If you don't like this model, you can disable it as a whole with the `ENTT_NO_ETO` definition. In this case, empty types are also instantiated and returned as if they were regular types.

**Comment by khalv** [comment]
All right, that sounds just like i'd want it to work! I was referring to this part:

https://github.com/skypjack/entt/wiki/Crash-Course:-entity-component-system#empty-type-optimization

_There doesn't exist a way to iterate empty types. Views and groups will never return instances of empty types_

But lower down, a line mentions how iterations are faster _because only the entities to which the type is assigned are considered_. I guess knowing the context, it's more clear, but i want to understand the phrase "iterating over a type" as "iterating over a list of all entities with a given type", and in that case, the first sentence makes it sound as if this is not possible. I think adding your example along with some introduction to the utility of empty types would improve that part of the documentation.

Thank you for taking your time to reply!

**Reply by skypjack** [reply]
You're welcome and thanks for the feedback. I'm biased when I read the documentation, so your suggestion is valuable! 🙂
